core/trino-main/src/main/java/io/trino/testing/TestingAccessControlManager.java

```diff
@@ -258,11 +276,22 @@ public Set<SchemaTableName> filterTables(SecurityContext context, String catalog
     @Override
     public void checkCanImpersonateUser(Identity identity, String userName)
     {
-        if (shouldDenyPrivilege(userName, userName, IMPERSONATE_USER)) {
-            denyImpersonateUser(identity.getUser(), userName);
+        // if allowImpersonationRoles is empty default behaviour is used
+        if (allowImpersonationRoles.isPresent()) {
+            boolean impersonationAllowed = identity.getEnabledRoles()
+                    .stream()
+                    .anyMatch(role -> allowImpersonationRoles.get().contains(role));
+            if (!impersonationAllowed) {
+                denyImpersonateUser(identity.getUser(), userName);
+            }
         }
-        if (denyPrivileges.isEmpty()) {
-            super.checkCanImpersonateUser(identity, userName);
+        else {
```

Comment:
remove the `else`, simply `return` when `allowImpersonationRoles` is not empty.

Reply:
No, that's the point this else is needed for other checks. There are 2 mechanisms that can denied impersonation:
- role based - the one I added, above this else
- privilege base  - the one inside this else - this one is preexisting